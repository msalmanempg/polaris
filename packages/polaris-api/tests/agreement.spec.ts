import {
  Agreement,
  AgreementCustomer,
  Customer,
  prisma,
  Prisma,
  Project,
  Unit,
} from "@polaris/prisma";

import {
  agreementCustomers,
  agreementProject,
  agreements,
  agreementUnit,
  customers,
  projects,
  rawAgreement,
  rawCustomers,
  rawNominees,
  rawSalesPersons,
  units,
} from "./support/fixtures/agreements";
import { testExpressApp } from "./support/requestFactory";

describe("agreements", () => {
  let createdProject: Project;
  let createdUnits: Unit[];
  let createdCustomers: Customer[];
  let createdAgreements: Agreement[];
  let createdAgreementCustomers: AgreementCustomer[];

  beforeEach(async () => {
    const [project] = projects;

    createdProject = await prisma.project.create({
      data: {
        ...(project as any),
      },
    });

    createdUnits = await Promise.all(
      units.map((unit) => {
        const unitData = {
          ...unit,
          unitNumber: `${createdProject.name} - ${unit.unitNumber as string}`,
          Project: {
            connect: {
              id: createdProject.id,
            },
          },
        } as any;

        return prisma.unit.create({ data: unitData });
      })
    );

    createdCustomers = await Promise.all(
      customers.map((customer) => {
        return prisma.customer.create({
          data: customer as any,
        });
      })
    );

    createdAgreements = await Promise.all(
      agreements.map((agreement, index) => {
        const agreementData: Prisma.AgreementCreateInput = {
          ...agreement,
          Unit: {
            connect: {
              id: createdUnits[index].id,
            },
          },
        } as any;

        return prisma.agreement.create({ data: agreementData });
      })
    );

    const [createdAgreement] = createdAgreements;

    createdAgreementCustomers = await Promise.all(
      agreementCustomers.slice(0, 2).map((agreementCustomer, index) => {
        const agreementCustomerData: Prisma.AgreementCustomerCreateInput = {
          ...agreementCustomer,
          Customer: {
            connect: {
              id: createdCustomers[index].id,
            },
          },
          Agreement: {
            connect: {
              id: createdAgreement.id,
            },
          },
        } as any;

        return prisma.agreementCustomer.create({ data: agreementCustomerData });
      })
    );
  });

  afterEach(async () => {
    const deleteAgreementCustomer = prisma.agreementCustomer.deleteMany({
      where: {
        id: {
          in: createdAgreementCustomers.map((agreementCustomer) => agreementCustomer.id),
        },
      },
    });

    const deleteCustomer = prisma.customer.deleteMany({
      where: {
        id: {
          in: createdCustomers.map((customer) => customer.id),
        },
      },
    });

    const deleteAgreement = prisma.agreement.deleteMany({
      where: {
        id: {
          in: createdAgreements.map((agreement) => agreement.id),
        },
      },
    });

    const deleteUnit = prisma.unit.deleteMany({
      where: {
        projectId: {
          in: createdProject.id,
        },
      },
    });

    const deleteProject = prisma.project.deleteMany({
      where: {
        id: {
          equals: createdProject.id,
        },
      },
    });

    await prisma.$transaction([
      deleteAgreementCustomer,
      deleteCustomer,
      deleteAgreement,
      deleteUnit,
      deleteProject,
    ]);
  });

  [
    {
      description: "with no filter",
      parameters: {
        pageOffset: 0,
        pageSize: 10,
      },
    },
    {
      description: "with status",
      parameters: {
        pageOffset: 0,
        pageSize: 10,
        status: "active",
      },
    },
  ].map(({ description, parameters }) => {
    it(`Should list all agreements ${description}`, async () => {
      await listAllAgreements(parameters);
    });
  });

  it("Should list all agreements for a project id", async () => {
    const parameters = {
      pageOffset: 0,
      pageSize: 10,
      projectId: createdProject.id,
    };

    await listAllAgreements(parameters);
  });

  it("Should list all agreements with stats", async () => {
    const parameters = {
      pageOffset: 0,
      pageSize: 10,
      includeStats: true,
    };

    const { body } = await testExpressApp().get("/api/agreement").query(parameters).expect(200);

    body.stats.sort((a: { status: string; count: number }, b: { status: string; count: number }) =>
      a.status.localeCompare(b.status)
    );

    expect(body).toMatchSnapshot({
      agreements: expect.any(Array),
      count: expect.any(Number),
    });
  });

  it("Should list all agreements for customer govt id", async () => {
    const parameters = {
      pageOffset: 0,
      pageSize: 10,
      customerGovtId: createdCustomers[0].govtId,
    };

    await listAllAgreements(parameters);
  });

  it("Should create new agreement", async () => {
    const project = await prisma.project.create({
      data: {
        ...(agreementProject as any),
        units: {
          createMany: {
            data: [agreementUnit],
          },
        },
      },
    });
    const rawUnit = await prisma.unit.findFirst({ where: { projectId: project.id } });
    const agreementData = {
      ...rawAgreement,
      unitId: rawUnit?.id,
      customers: rawCustomers,
      nominees: rawNominees,
      salesPerson: rawSalesPersons,
    };

    const { body } = await testExpressApp()
      .post("/api/agreement/add")
      .send(agreementData)
      .expect(201);
    expect(body).toMatchSnapshot(agreementSnapshotMatch(body));
  });

  it("should check availability of an agreement number", async () => {
    const agreementNumber = "XYZ0123456789";

    const { body } = await testExpressApp()
      .get(`/api/agreement/check_agreement_number/${agreementNumber}`)
      .expect(200);
    expect(body).toEqual(true);
    expect(body).toMatchSnapshot();
  });

  it("Should fecth details of an agreement", async () => {
    const agreement = await prisma.agreement.findFirst({
      where: { agreementNumber: agreements[0].agreementNumber },
    });

    const { body } = await testExpressApp().get(
      `/api/agreement/detail/${agreement ? agreement.id : 0}`
    );
    expect(body).toMatchSnapshot(agreementSnapshotMatch(body));
  });
});

const listAllAgreements = async (parameters: any) => {
  const { body } = await testExpressApp().get("/api/agreement/all").query(parameters).expect(200);

  expect(body).toMatchSnapshot({
    agreements: expect.any(Array),
  });

  const agreements = body.agreements.sort((a: { id: number }, b: { id: number }) => a.id < b.id);
  matchAgreements(agreements);
};

const matchAgreements = (agreements: any) => {
  agreements.map((agreement: any) => {
    expect(agreement).toMatchSnapshot({
      ...agreement,
      id: expect.any(Number),
      unit: {
        id: expect.any(Number),
        project: {
          id: expect.any(Number),
        },
      },
      customer: expect.any(Object),
    });
  });
};

const agreementSnapshotMatch = (body: any) => {
  return {
    id: expect.any(Number),
    createdAt: expect.any(String),
    updatedAt: expect.any(String),
    unitId: expect.any(Number),
    Unit: {
      ...body.Unit,
      id: expect.any(Number),
      createdAt: expect.any(String),
      updatedAt: expect.any(String),
      projectId: expect.any(Number),
    },
    customers: body.customers.map(
      (customer: any) =>
        ({
          ...customer,
          id: expect.any(Number),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
          agreementId: expect.any(Number),
          customerId: expect.any(Number),
        } as unknown)
    ),
    nominees: body.nominees.map(
      (nominee: any) =>
        ({
          ...nominee,
          id: expect.any(Number),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
          agreementId: expect.any(Number),
        } as unknown)
    ),
    salesPerson: body.salesPerson.map(
      (salePerson: any) =>
        ({
          ...salePerson,
          id: expect.any(Number),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
          agreementId: expect.any(Number),
        } as unknown)
    ),
  };
};
