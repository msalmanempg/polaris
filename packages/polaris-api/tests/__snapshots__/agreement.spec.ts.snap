// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agreements Should create new agreement 1`] = `
Object {
  "Unit": Object {
    "basePrice": 5000000,
    "bed": 3,
    "completionDate": "2021-02-19T05:20:07.269Z",
    "createdAt": Any<String>,
    "grossArea": 1250,
    "id": Any<Number>,
    "location": "Block A, Street 5",
    "meta": null,
    "netArea": 1250,
    "projectId": Any<Number>,
    "publishedPrice": 5500000,
    "status": "booked",
    "type": "residential",
    "unitNumber": "unit-agreement",
    "updatedAt": Any<String>,
  },
  "agreementDate": "2022-02-15T00:03:58.000Z",
  "agreementNumber": "ABC101019",
  "bookingDate": "2022-02-15T00:03:58.000Z",
  "bookingPrice": 10000,
  "company": Object {},
  "createdAt": Any<String>,
  "customers": Array [
    Object {
      "address": "Some Address",
      "agreementId": Any<Number>,
      "createdAt": Any<String>,
      "customerId": Any<Number>,
      "email": "m.ali@gmail.com",
      "id": Any<Number>,
      "isPrimary": true,
      "meta": null,
      "primaryPhoneNumber": "+923006898283",
      "secondaryPhoneNumber": null,
      "type": "company",
      "updatedAt": Any<String>,
    },
  ],
  "discountValue": 100,
  "id": Any<Number>,
  "installmentEndDate": "2022-02-15T00:03:58.000Z",
  "installmentStartDate": "2022-02-15T00:03:58.000Z",
  "leadId": 101019,
  "nominees": Array [
    Object {
      "agreementId": Any<Number>,
      "contactNumber": "03216786543",
      "createdAt": Any<String>,
      "email": "arslan.ali@gmail.com",
      "fullName": "Arslan Ali",
      "govtId": "10203040506070",
      "govtIdType": "cnic",
      "id": Any<Number>,
      "meta": null,
      "nomineeFor": "Muhammad Ali",
      "relationship": "Brother",
      "updatedAt": Any<String>,
    },
  ],
  "notarisedDate": "2022-02-15T00:03:58.000Z",
  "posessionValue": 2000,
  "publishedPrice": 10000,
  "salesPerson": Array [
    Object {
      "agreementId": Any<Number>,
      "createdAt": Any<String>,
      "designation": "Manager",
      "email": "ali.raza@gmail.com",
      "fullName": "Ali Raza",
      "govtId": "19293949596979",
      "id": Any<Number>,
      "phoneNumber": "03136799098",
      "sortLevel": 0,
      "updatedAt": Any<String>,
    },
  ],
  "status": "active",
  "unitId": Any<Number>,
  "updatedAt": Any<String>,
}
`;

exports[`agreements Should fecth details of an agreement 1`] = `
Object {
  "Unit": Object {
    "basePrice": 5000000,
    "bed": 3,
    "completionDate": "2021-02-19T05:20:07.269Z",
    "createdAt": Any<String>,
    "grossArea": 1250,
    "id": Any<Number>,
    "location": "Block A, Street 5",
    "meta": null,
    "netArea": 1250,
    "projectId": Any<Number>,
    "publishedPrice": 5500000,
    "status": "available",
    "type": "residential",
    "unitNumber": "Project 1 for agreements - unit-a for agreements",
    "updatedAt": Any<String>,
  },
  "agreementDate": "2021-02-22T05:20:07.269Z",
  "agreementNumber": "AG-01",
  "bookingDate": "2021-02-20T05:20:07.269Z",
  "bookingPrice": 5000000,
  "createdAt": Any<String>,
  "customers": Array [
    Object {
      "address": "Lahore, Pakistan",
      "agreementId": Any<Number>,
      "createdAt": Any<String>,
      "customerId": Any<Number>,
      "email": "individual@email.com",
      "id": Any<Number>,
      "isPrimary": true,
      "meta": null,
      "primaryPhoneNumber": "9230050000000",
      "secondaryPhoneNumber": "9230050000001",
      "type": "individual",
      "updatedAt": Any<String>,
    },
    Object {
      "address": "Lahore, Pakistan",
      "agreementId": Any<Number>,
      "createdAt": Any<String>,
      "customerId": Any<Number>,
      "email": "individual2@email.com",
      "id": Any<Number>,
      "isPrimary": false,
      "meta": null,
      "primaryPhoneNumber": "9230040000000",
      "secondaryPhoneNumber": "9230040000001",
      "type": "individual",
      "updatedAt": Any<String>,
    },
  ],
  "discountValue": 500000,
  "id": Any<Number>,
  "installmentEndDate": "2022-02-23T05:20:07.269Z",
  "installmentStartDate": "2021-02-23T05:20:07.269Z",
  "leadId": 1,
  "nominees": Array [],
  "notarisedDate": "2021-02-21T05:20:07.269Z",
  "posessionValue": 1000000,
  "publishedPrice": 5500000,
  "salesPerson": Array [],
  "status": "active",
  "unitId": Any<Number>,
  "updatedAt": Any<String>,
}
`;

exports[`agreements Should list all agreements for a project id 1`] = `
Object {
  "agreements": Any<Array>,
  "count": 2,
  "pageOffset": 0,
  "pageSize": 10,
}
`;

exports[`agreements Should list all agreements for a project id 2`] = `
Object {
  "agreementDate": "2021-02-22T05:20:07.269Z",
  "agreementNumber": "AG-01",
  "customer": Any<Object>,
  "id": Any<Number>,
  "noticesCount": 0,
  "overduePayments": 0,
  "paymentPercentage": 0,
  "status": "active",
  "unit": Object {
    "id": Any<Number>,
    "project": Object {
      "id": Any<Number>,
      "name": "Project 1 for agreements",
    },
    "unitNumber": "Project 1 for agreements - unit-a for agreements",
  },
}
`;

exports[`agreements Should list all agreements for a project id 3`] = `
Object {
  "agreementDate": "2021-02-22T05:20:07.269Z",
  "agreementNumber": "AG-02",
  "customer": Any<Object>,
  "id": Any<Number>,
  "noticesCount": 0,
  "overduePayments": 0,
  "paymentPercentage": 0,
  "status": "draft",
  "unit": Object {
    "id": Any<Number>,
    "project": Object {
      "id": Any<Number>,
      "name": "Project 1 for agreements",
    },
    "unitNumber": "Project 1 for agreements - unit-b for agreements",
  },
}
`;

exports[`agreements Should list all agreements for customer govt id 1`] = `
Object {
  "agreements": Any<Array>,
  "count": 1,
  "pageOffset": 0,
  "pageSize": 10,
}
`;

exports[`agreements Should list all agreements for customer govt id 2`] = `
Object {
  "agreementDate": "2021-02-22T05:20:07.269Z",
  "agreementNumber": "AG-01",
  "customer": Any<Object>,
  "id": Any<Number>,
  "noticesCount": 0,
  "overduePayments": 0,
  "paymentPercentage": 0,
  "status": "active",
  "unit": Object {
    "id": Any<Number>,
    "project": Object {
      "id": Any<Number>,
      "name": "Project 1 for agreements",
    },
    "unitNumber": "Project 1 for agreements - unit-a for agreements",
  },
}
`;

exports[`agreements Should list all agreements with no filter 1`] = `
Object {
  "agreements": Any<Array>,
  "count": 2,
  "pageOffset": 0,
  "pageSize": 10,
}
`;

exports[`agreements Should list all agreements with no filter 2`] = `
Object {
  "agreementDate": "2021-02-22T05:20:07.269Z",
  "agreementNumber": "AG-01",
  "customer": Any<Object>,
  "id": Any<Number>,
  "noticesCount": 0,
  "overduePayments": 0,
  "paymentPercentage": 0,
  "status": "active",
  "unit": Object {
    "id": Any<Number>,
    "project": Object {
      "id": Any<Number>,
      "name": "Project 1 for agreements",
    },
    "unitNumber": "Project 1 for agreements - unit-a for agreements",
  },
}
`;

exports[`agreements Should list all agreements with no filter 3`] = `
Object {
  "agreementDate": "2021-02-22T05:20:07.269Z",
  "agreementNumber": "AG-02",
  "customer": Any<Object>,
  "id": Any<Number>,
  "noticesCount": 0,
  "overduePayments": 0,
  "paymentPercentage": 0,
  "status": "draft",
  "unit": Object {
    "id": Any<Number>,
    "project": Object {
      "id": Any<Number>,
      "name": "Project 1 for agreements",
    },
    "unitNumber": "Project 1 for agreements - unit-b for agreements",
  },
}
`;

exports[`agreements Should list all agreements with stats 1`] = `
Object {
  "agreements": Any<Array>,
  "count": Any<Number>,
  "pageOffset": 0,
  "pageSize": 10,
  "stats": Array [
    Object {
      "count": 1,
      "status": "active",
    },
    Object {
      "count": 2,
      "status": "all",
    },
    Object {
      "count": 1,
      "status": "draft",
    },
  ],
}
`;

exports[`agreements Should list all agreements with status 1`] = `
Object {
  "agreements": Any<Array>,
  "count": 1,
  "pageOffset": 0,
  "pageSize": 10,
}
`;

exports[`agreements Should list all agreements with status 2`] = `
Object {
  "agreementDate": "2021-02-22T05:20:07.269Z",
  "agreementNumber": "AG-01",
  "customer": Any<Object>,
  "id": Any<Number>,
  "noticesCount": 0,
  "overduePayments": 0,
  "paymentPercentage": 0,
  "status": "active",
  "unit": Object {
    "id": Any<Number>,
    "project": Object {
      "id": Any<Number>,
      "name": "Project 1 for agreements",
    },
    "unitNumber": "Project 1 for agreements - unit-a for agreements",
  },
}
`;

exports[`agreements should check availability of an agreement number 1`] = `true`;
